library/core/src/convert/mod.rs

```diff
@@ -121,9 +124,66 @@ pub const fn identity<T>(x: T) -> T {
 ///
 /// # Generic Implementations
 ///
-/// - `AsRef` auto-dereferences if the inner type is a reference or a mutable
-///   reference (e.g.: `foo.as_ref()` will work the same if `foo` has type
-///   `&mut Foo` or `&&mut Foo`)
+/// `AsRef` auto-dereferences if the inner type is a reference or a mutable reference
+/// (e.g.: `foo.as_ref()` will work the same if `foo` has type `&mut Foo` or `&&mut Foo`).
+///
+/// Note that due to historic reasons, the above currently does not hold generally for all
+/// [dereferenceable types], e.g. `foo.as_ref()` will *not* work the same as
+/// `Box::new(foo).as_ref()`. Instead, many smart pointers provide an `as_ref` implementation which
+/// simply returns a reference to the [pointed-to value] (but do not perform a cheap
+/// reference-to-reference conversion for that value). However, [`AsRef::as_ref`] should not be
+/// used for the sole purpose of dereferencing; instead ['`Deref` coercion'] can be used:
+///
+/// [dereferenceable types]: core::ops::Deref
+/// [pointed-to value]: core::ops::Deref::Target
+/// ['`Deref` coercion']: core::ops::Deref#more-on-deref-coercion
+///
+/// ```
+/// let x = Box::new(5i32);
+/// // Avoid this:
+/// // let y: &i32 = x.as_ref();
+/// // Better just write:
+/// let y: &i32 = &x;
+/// ```
+///
+/// Types which implement [`Deref`] should consider implementing `AsRef<T>` as follows:
+///
+/// [`Deref`]: core::ops::Deref
+///
+/// ```
+/// # use core::ops::Deref;
+/// # struct SomeType;
+/// # impl Deref for SomeType {
+/// #     type Target = [u8];
+/// #     fn deref(&self) -> &[u8] {
+/// #         &[]
+/// #     }
+/// # }
+/// impl<T> AsRef<T> for SomeType
+/// where
+///     T: ?Sized,
+///     <SomeType as Deref>::Target: AsRef<T>,
+/// {
+///     fn as_ref(&self) -> &T {
+///         self.deref().as_ref()
+///     }
+/// }
+/// ```
+///
+/// # Reflexivity
+///
+/// Ideally, `AsRef` would be reflexive, that is there is an `impl<T: ?Sized> AsRef<T> for T`, with
```

Comment:
I stumbled over this phrasing a couple times.
Do you think this conveys your meaning?

```text
Ideally, `AsRef` would be reflexive, with an `impl ...`
```

Reply:
I wanted to make clear that the `impl` is actually the meaning of (i.e. the same as) reflexivity in this context. This doesn't seem so clear with your proposed change (to me).

How about: Ideally, `AsRef` would be reflexive, i.e. with an `impl ...`

But note how there is a double "with" then, which also impedes reading flow a bit:

> Ideally, AsRef would be reflexive, [i.e.] **with** an impl<T: ?Sized> AsRef<T> for T, **with** [as_ref](https://doc.rust-lang.org/std/convert/trait.AsRef.html#tymethod.as_ref) simply returning its argument unchanged. […]

The second "with" could be replaced with a "where" (and then "returning" be replaced with "returns"), but I wanted to avoid using `impl` and "where" because it could be confused with the meaning of `where` as a keyword.

For these reasons, I personally like the original phrasing more (but I'm not a native English speaker). But maybe there is a better way to phrase this or the double "with" isn't that bad.

Perhaps "that is" could just be replaced with "i.e.":

> Ideally, AsRef would be reflexive, i.e. there is an impl<T: ?Sized> AsRef<T> for T, with [as_ref](https://doc.rust-lang.org/std/convert/trait.AsRef.html#tymethod.as_ref) simply returning its argument unchanged. […]

What do you think about that?

Reply:
I like the i.e, but "i.e. there is an impl" contradicts the "would be" hypothetical. It makes it sound like the impl already does exist.

How about just "i.e. `impl ...`, with ...", with no "there is"?

Reply:
> I like the i.e, but "i.e. there is an impl" contradicts the "would be" hypothetical.

Ah, I see now why you stumbled upon it, and I agree it's not optimally phrased. It would need to be "there was" or "there would be". Your proposal would be:

> Ideally, `AsRef` would be reflexive, i.e. `impl<T: ?Sized> AsRef<T> for T`, with [`as_ref`](https://doc.rust-lang.org/std/convert/trait.AsRef.html#tymethod.as_ref) simply returning its argument unchanged.

I would rather write:

> Ideally, `AsRef` would be reflexive, i.e. there would be an `impl<T: ?Sized> AsRef<T> for T`, with [`as_ref`](https://doc.rust-lang.org/std/convert/trait.AsRef.html#tymethod.as_ref) simply returning its argument unchanged.

I slightly prefer the second (longer) version because there may be other `impl`s too. The first (shorter) version feels slightly confusing in that matter.

Reply:
I went for the second (longer) version and decided to remove the comma before "with" (I think it's superfluous). So it now reads:

> Ideally, `AsRef` would be reflexive, i.e. there would be an `impl<T: ?Sized> AsRef<T> for T` with [`as_ref`](https://doc.rust-lang.org/std/convert/trait.AsRef.html#tymethod.as_ref) simply returning its argument unchanged.

I updated the PR with e6b761b902fe80ba454368d078fb834554981109. Let me know if you still think there is potential to make it better (or if you think that change made it worse).

Reply:
LGTM 👌